File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const pairs = {
        ')': '(',
        '}': '{',
        ']': '[',
    };

    const opening = new Set(Object.values(pairs));
    const stack = [];

    for (const c of s) {
        if (opening.has(c)) {
            stack.push(c);
        } else if (c in pairs) {
            if (stack.length === 0 || stack.pop() !== pairs[c]) {
                return false;
            }
        } else {
            // Optional: handle unexpected characters
            return false;
        }
    }

    return stack.length === 0;
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {')': '(', '}': '{', ']': '['}
        stack = []

        for c in s:
            if c in pairs.values():  # Opening brackets
                stack.append(c)
            elif c in pairs:  # Closing brackets
                if not stack or stack.pop() != pairs[c]:
                    return False
            else:
                # Optional: handle invalid characters if needed
                return False

        return not stack
